### DISCUSSION TITLE
Creating a model.class for structured (header-)element content

### DISCUSSION THREAD

**Discussion by riedde** [music-encoding/metadata-ig#20] [Ideas]
The unsatisfying fact that the content of multiple header elements ha too many elements must be solved some time. In the IG-meeting on April 14th we agreed to work on this topic after the MEC 2022.

The goal is to define models that contain a senseful group of elements that can be proposed as element content for several metadata elements. (e.g., it is  unsatisfying that `<foliation>` may contain `<watermark>`)

The ODD-Team should be included right from the beginning!

**Comment by ahankinson** [comment]
Not to be a rainy cloud, but I would caution against large-scale restructuring efforts. I am only one voice here, so please weigh my comments accordingly. But I think it's important to understand why these are the way they are before we commit to doing a full-scale re-work [1]. So if you'll permit me to comment on this proposal...

TEI, and by extension those parts of MEI that are based on TEI, is not a 'traditional' metadata scheme as you might expect if you're thinking about metadata from the MARC / Dublin Core tradition, or even the 3x5 catalogue card and ISBD tradition. In these metadata schemes, you have a field and a value with fairly precise, and concise, contents. So if you're thinking about metadata with a view towards making it align with these formats, you might expect an element like  `<foliation>` to contain a highly restricted, atomic set of values, and be a simple statement such as `32 folios`.

However, if you [read the TEI guidelines](https://tei-c.org/release/doc/tei-p5-doc/en/html/MS.html#msphfo) and observe how TEI is used "in the real world" you will see that the contents of elements like `<foliation>` are not simple atomic values, but can actually be highly-specialized text markup. Real-world use can look more like this:

```xml
<foliation>
A pencil foliation (s. xx) assigns the flyleaf the number ii with the frontispiece 
attached to it (see below) as fol. i; for the main body, two pen foliation in upper right 
corners, the chronologically earlier being that provided by Sir Walter Cope, and the 
second added by the time of binding for Robert Cotton (on both these collectors, 
see <hi rend="smallcaps">provenance</hi>). The earlier foliation runs fols 9–362, 
but with the last leaves, from fol. 355, having been originally 1–8; fol. 289–93 precede 
the folios Cope numbered 286–88, which the Cottonian foliation renumbers as 294–96, 
with the original 294–96 also renumbered 297–99; there are also errors, with fol. 197 
repeated, 203 omitted and an unnumbered leaf after fol. 218.
</foliation>
```

https://github.com/bodleian/medieval-mss/blob/d231112115d53e52edcfcacaf6354e61885071a6/collections/Christ_Church/Christ_Church_MS_343.xml#L97

In these cases, you can begin to see how elements such as `<watermark>` might actually be useful, as well as `<person>`, `<locus>`, or several other elements that at first glance do not directly relate to the foliation. So the purpose of these elements isn't to create a simple atomic "key and value" record, like you might expect in a library record, but to actually allow for marked-up, annotated descriptive text more in line with the tradition of manuscript scholarship and descriptive cataloguing. Of course, you *can* have more atomic values, but you also have the tools to expand on this if your use-case requires it.

To give an example of what I mean more broadly, here are two pages from the 19th/20th century Summary Catalogue of Manuscripts in the Bodleian describing MSS. Hatton 113-14. I give this as a sort of "origin story" for where this tradition comes from:

https://medieval.bodleian.ox.ac.uk/images/ms/aaq/aaq0326.gif
https://medieval.bodleian.ox.ac.uk/images/ms/aaq/aaq0327.gif

This entry is designed as a brief scholarly description, in prose, of the manuscript in question. What it is not, though, is a simple "bibliographic" record, as we might expect from a later library catalogue. [2]

Looking at the newer electronic record of the same: 

https://medieval.bodleian.ox.ac.uk/catalog/manuscript_6027

and the corresponding TEI file:

https://github.com/bodleian/medieval-mss/blob/master/collections/Hatton/MSS_Hatton_113-14.xml

We can see that this still follows the tradition of the 19th century catalogues, despite being translated into a digital context. Again, these are highly descriptive, but fundamentally prose-formatted, catalogues. Accordingly, these records are meant to be processed more like a highly-descriptive HTML markup, rather than as atomic metadata values.

Perhaps the next question, however, is that if we're MEI, why are we following the TEI at all? We have a collective decision as a community that MEI is not TEI, and so we can diverge from that specification where we need to. To be clear: I agree with the fundamentals of this sentiment. Music has needs that are different than text, and we need to have freedom to diverge when we see fit.

But I think this is tempered by three factors:

1. The world has quite a few "key/value" metadata formats already, and probably does not need another. This is also true in the inverse: There are very few formats that can be used to do "scholarly catalogue markup" using prose -- TEI being the most widely used. EAD might, in some cases, also be considered for this, but it is designed more for electronic finding aids, rather than catalogues. There are no formats I am aware of that provide this for music documents, other than MEI.
2. Aligning ourselves with TEI makes it much easier for scholars to make the jump between the two formats, should the need arise. We then have access to a larger community, and that community has access to ours, and both communities benefit. More divergence means higher barriers to entry for those who have learned TEI but then want to apply the same principles to music. TEI is taught as a fundamental skill in many Digital Humanities classes, while MEI is a much more specialized format and so is often left to individual scholars to pick up on their own, if they feel it meets their needs.
3. It's important to recognize that we gain the thought that went into the modelling of such elements from the TEI community. This is not a trivial process; the `msDesc` section took over a decade, if not almost two, to be formulated and to 'catch on'. Re-modelling takes serious effort and wide-spread consultation, and if we can use the work having been done already, this is a net benefit to the MEI effort.

I say all of this in the hopes that I am not discouraging the work of the IG, nor trivializing their discussions, but also in the spirit of providing some context to "why things are the way they are" and potentially filling some gaps in the community knowledge base. Nor am I saying that things are perfect -- I think there are likely areas that can be improved.

So I have three basic concerns with this proposal:

1. Standards are not software, and the creation of standards is fundamentally more political than technical. Depending on the scale of re-work that is envisioned, re-modelling takes a lot of dedicated effort, consultation, and communication. I expect that such an effort, should it go forward, would take at least a year; in my experience, probably more, before any actual code in the spec changes. It is also very important for the IG to recognize that, while their meetings are open and anyone can join, not all community members can spare the time to participate. I say this only to caution that, while discussions in the IGs meetings may be helpful to those in the IG, it is very important that the outcomes of these discussions are tabled in wider venues prior to the changes being proposed, and wider community consensus sought. To put it more succinctly: The Pull Requests to make the change should only come in after there is wider community acceptance and consensus of the direction of change. The bigger the remodelling, the more important the communication.
2. We should recognize that we would lose a significant portion of cross-compatibility with TEI should we re-model this section. As I said above, fundamentally I'm OK with diverging if we need to, but the IG should first make the case that there is a need to diverge. I believe this will require understanding the goals of the TEI efforts itself, and then making the case that this does not suit the needs of the MEI.
3. At a more abstract level, having things like "watermark" being available on "foliation" doesn't actually *hurt* anything. It is not a bug, and for some it may be a feature. It may not fit within a narrow definition of what foliation is, but as discussed above, the divergence in understanding of the content of these elements may be issues in understanding how metadata in MEI, and by extension TEI, is framed. Frankly, I think this is primarily because of lack of documentation than any specific modelling problems.

To close, I want to say "Thank you" for posting this note and making it a discussion well in advance of the change. This is a very useful and open way of making sure a wide range of voices are heard, and so that we can arrive at a solution that benefits everyone.

---

[1] Some may wonder why I have opinions about this stuff. My experience in the metadata section of the TEI is based on how it is being used in institutions in the UK to create large manuscript catalogues. I was the system architect for the Fihrist Union Catalogue of Islamic Manuscripts (https://www.fihrist.org.uk), the Medieval Manuscripts at Oxford (https://medieval.bodleian.ox.ac.uk) and six other catalogues at Oxford. I also worked closely with colleagues who were framing and publishing the Encoding Guidelines for Manuscript Description (https://msdesc.github.io/consolidated-tei-schema/msdesc.html) and consulted on data processing and discoverability of the TEI records. I say this not to boast, but to provide some background in my experience with this particular section of the TEI.

[2] As a side note, it is interesting to consider these traditions in terms of the ability to access the materials and their intended use. A catalogue card is, fundamentally, about finding mass-produced books on a shelf; I believe MARC, Dublin Core, METS, etc. are extensions of this use-case. A descriptive catalogue is more aligned towards scholarship where access to the material is restricted, either because it is in a distant place, or because it is rare or fragile that only certain people can access it. In these cases, having prose descriptions that are widely distributed are much more useful since they serve as a surrogate to the material, and a summary of the history and scholarship of the item to date in a way that does not translate easily to breaking it down into atomic fields and values. To take the `<foliation>` example, you may only know how an item is foliated, or you may actually need to describe multiple systems of foliation in use, and when they were added, and what was used to add them (pen, ink), whether there are mistakes in the foliation, etc.

**Reply by riedde** [reply]
A very helpful comment! And many thanks for the examples!

> I say all of this in the hopes that I am not discouraging the work of the IG, nor trivializing their discussions, but also in the spirit of providing some context to "why things are the way they are" and potentially filling some gaps in the community knowledge base. Nor am I saying that things are perfect -- I think there are likely areas that can be improved.

I'm sure I can speak for the ig when I say that we are thankful for filling gaps in our knowledge. We often ask our self why things are as they are.

To prose text: I'm sorry if my first notes where rudimentary, so I try to clarify them a bit: The question is not if 'prose' or 'atomic values' should be the absolute way of encoding. Both are used and both have their right to exist. The question, for me, is if we can extent/adjust the way of encoding metadata as 'atmoic values' that it is more clear and easier to use in general - as an additional option.

The way it is right now it is a model that is always mixed in any case of use. So my idea is to discuss here if it could be possible to create a model class, e.g., for use when encoding structured metadata, not prose, that could be used alternatively. As I said, it is not my idea to restrict everything to one usecase. I'm thinking about extending the running system for a special purpose. So if there will be no break up with the prose encoding (which is very improtant as you pointed out!) or backwards compatibility, in my eyes there is no strong argument not to talk about. (It is a discussion, not yet an issue.)

> I say this only to caution that, while discussions in the IGs meetings may be helpful to those in the IG, it is very important that the outcomes of these discussions are tabled in wider venues prior to the changes being proposed, and wider community consensus sought. To put it more succinctly: The Pull Requests to make the change should only come in after there is wider community acceptance and consensus of the direction of change. The bigger the remodelling, the more important the communication. 

Thanks for that hint. We tried some years ago to bring up discussions and proposals to other groups like the ODD-Team so they could find ways to integrate our 'ideas and wishes' into the schema. This procedure was not working fine and the ig was told to model thinks by themselve and to open a pull request. Since there is no defined workflow maybe a basic one should be defined by the board (I think an issue for that is already opened). We will try to make our ideas, discussions, issues and proposals public. But I agree that more communication is needed in any way!

1. Concerning the fact that not everybody is able to join our meetings: We didn't get any notice that people outside the ig are interested in being a part of what we are doing for the last year, as I know. We support participation! Because of this we brought our issues and discussions to the github universe so anybody is able to participate even if the meetings can't be attended.  
2. Concerning compatibility with TEI: I understand your point of view and I can also agree on many aspects. But if compatibility is so important, why does TEI have, e.g., objectDesc as a wrapper for layoutDesc and supportDesc and MEI not? Wouldn't it be more clear what supportDesc means if it is related to an object? And there are some other things that are different that would make using both standards easier if they do not differ for a non obvious reason (e.g., TEI: orgName, MEI: corpName)?

3. It hurts me (but I know that this is no argument). But it does not break anything. The more important thing, for me, is the lack of documentation. The ig tries to fix that but sometimes we come to a point where the question is asked why not the people how know much more of the 'implicit' functionality of MEI do that job (or have done it before). I do not want to point out how is guilty, because that is not purposeful at all! But maybe we can communicate better what we are doing or willing to do in the future.

>and so that we can arrive at a solution that benefits everyone.

This is my and our intention :-)

**Comment by riedde** [comment]
An other thing that I can imagine is to bring up an official mei customization for encoding structured metadata maintained by the @music-encoding/ig-metadata

**Comment by kepper** [comment]
This is a very important discussion, important way beyond the issue at hand. I don't want to go into details here, but would like to emphasize some of the arguments in here. First, and most importantly: work on standards is mostly a communication thing. Having a good technical solution for a problem at hand is certainly important, but explaining it to others involved is at least as important. The bigger the proposal – be it something new or a revision of existing parts – the bigger the need for communication – within any given IG, but also beyond. 
At the same time, such discussions take away a considerable amount of time, time in which no technical progress is made. This is frustrating, especially when that time is lacking when it comes to writing documentation (a truely sisyphean task…). However, this is the price of working on a _widely accepted_ standard, and ideally, those discussions will lay the foundation for better documentation, as with Andrew's wonderful example from above. 
I think we all agree that commonalities between MEI and TEI are a useful default, and that we surely need to reserve the freedom to diverge from TEI when we have good arguments. However, the decade-long work on a consensus that Andrew brought up isn't really finished on either end – like MEI, the TEI is also a moving target. Sometimes, differences between MEI and TEI originate in continued discussions on TEI's side that went unnoticed (or unresponded) on the MEI side. Eventually, this is just the same effect as with the work of the IGs within MEI – communication hasn't reached everyone who will be affected. This is also not to blame anyone, it just shows how complicated work on standards is. We should definitely try to be more involved in TEI's processes, but this will take away even more time…
All that said, I think that the last proposal brought up by Dennis is a good one – before actually changing the standard, a reasonable first step could be to identify a _good_ practice on how to use it. If eventually adopted / accepted by the wider community, it will turn into a _best practice recommendation_. Evaluating the use of such a recommendation will eventually give arguments to modify the underlying standard. Such a proposal, going along with a clearly defined use-case, would not claim to be authoritative for all of MEI, and would thus not "threat" other uses of the corresponding parts of MEI in the same way a rewrite would. Following this approach could therefore be a first step into the original direction, without triggering the same kind of problems.

**Comment by pe-ro** [comment]
Here's a compromise solution for allowing structured or "atomic" content while also providing additional guidance:
```
<define name="mei_captureMode">
    <element name="captureMode">
      <a:documentation xmlns:a="http://relaxng.org/ns/compatibility/annotations/1.0"
        >(capture mode) The means used to record notation, sound, or images in the
        production of a source/manifestation (e.g., analogue, acoustic, electric, digital,
        optical etc.).</a:documentation>
      <choice>
        <optional>
          <element name="term">
            <ref name="mei_att.common.attributes"/>
            <ref name="mei_att.authorized.attributes"/>
            <ref name="mei_att.dataPointing.attributes"/>
            <ref name="mei_att.lang.attributes"/>
            <choice>
              <value>acoustical capture, direct storage</value>
              <value>electrical capture, analog electrical storage</value>
              <value>electrical capture, digital storage</value>
              <value>electrical capture, direct storage</value>
              <text/>
            </choice>
          </element>
        </optional>
        <group>
          <zeroOrMore>
            <ref name="mei_head"/>
          </zeroOrMore>
          <zeroOrMore>
            <ref name="mei_p"/>
          </zeroOrMore>
        </group>
      </choice>
      <ref name="mei_att.common.attributes"/>
      <ref name="mei_att.authorized.attributes"/>
      <ref name="mei_att.bibl.attributes"/>
      <ref name="mei_att.lang.attributes"/>
      <empty/>
    </element>
  </define>
```
In this approach, the `<term>` element is defined contextually within `<captureMode>` so that either a suggested value or arbitrary text can be used. `<term>` is made a member of `att.authorized` so that the term text can be linked to an authoritative source.  Just to be clear, `<term>` allows only text without all of the sub-elements allowed by `textPhraseLike.limited'.  However, I have also been thinking about a new  model for phrase-level text along the lines of 
```
<define name="mei_model.textPhraseLike.basic">
  <choice>
    <ref name="mei_model.annotLike"/>
    <ref name="mei_model.figureLike"/>
    <ref name="mei_model.locrefLike"/>
    <ref name="mei_model.rendLike"/>
    <ref name="mei_seg"/>
    <ref name="mei_symbol"/>
  </choice>
</define>
```
that might be useful in "atomic" text.

Using the proposed model of `<captureMode>`, all of the following are allowed:
```
<captureMode>
  <head/>
  <p/>
</captureMode>
<captureMode>
  <term auth="MARC" auth.uri="http://id.loc.gov/vocabulary/mcapturestorage/acds">acoustical capture, direct storage</term>
</captureMode>
<captureMode>
  <!-- A term can be used that's neither one of the suggested values nor recognized by another authority. -->
  <term>foobar</term>
</captureMode>
```
In oXygen, but probably not elsewhere, the suggested values for `<term>` are provided when one types the trailing '>' of `<term>`. In any case, the proposed model allows some documentation of the model's expectations in the schema itself, which eases some of the burden of creating supporting text in the Guidelines.

Unfortunately, Pure ODD doesn't allow this kind of construction; that is, `<elementSpec>` cannot contain `<elementSpec>`.  So, we'd either have to use a macroSpec for each re-definition of `<term>` or resort to ODD that relies on a block of RNG ("impure" ODD?). For the sake of consistency, we'll probably also have to review, and likely modify, any element that uses the struc-unstrucContent macro.

At the same time, I think it would be wise to 
1. make `<term>` a member of `att.authorized`,
2. remove membership in `att.bibl`  from `<term>`, and
3. remove `@sync` (and perhaps all the members of `att.linking`) from metadata elements 
that occur in `<physDesc>` (and perhaps other elements that occur only/primarily in the 
header).

I doubt that anyone has yet used any of the attributes in `att.linking` on metadata elements or `@analog` on `<term>`, so it's probably not worth documenting their deprecation first and actually removing them later.  According to the old saw, it's better to ask for forgiveness than permission. :-)

I won't be offended if the consensus is that this is too much work or is going too far because it breaks the current schema.  But, my attitude has always been that the schema isn't holy and that breaking it to allow for better/more precise markup is permissible, provided there's a reasonable upgrade path from one version to the next.

**Comment by pe-ro** [comment]
This version is simpler still.  The choice block containing textual values can be implemented as a macro, making it easier to modify later.  It also avoids the issue of redefining `<term>` contextually:
```
<define name="mei_captureMode">
    <element name="captureMode">
      <a:documentation xmlns:a="http://relaxng.org/ns/compatibility/annotations/1.0"
        >(capture mode) The means used to record notation, sound, or images in the
        production of a source/manifestation (e.g., analogue, acoustic, electric, digital,
        optical etc.).</a:documentation>
      <choice>
        <group>
          <ref name="mei_att.authorized.attributes"/>
          <ref name="mei_att.bibl.attributes"/>
          <choice>
          <value>acoustical capture, direct storage</value>
          <value>electrical capture, analog electrical storage</value>
          <value>electrical capture, digital storage</value>
          <value>electrical capture, direct storage</value>
          <text/>
        </choice></group>
        <group>
          <zeroOrMore>
            <ref name="mei_head"/>
          </zeroOrMore>
          <zeroOrMore>
            <ref name="mei_p"/>
          </zeroOrMore>
        </group>
      </choice>
      <ref name="mei_att.common.attributes"/>
      <!--<ref name="mei_att.authorized.attributes"/>-->
      <!--<ref name="mei_att.bibl.attributes"/>-->
      <ref name="mei_att.lang.attributes"/>
      <empty/>
    </element>
  </define>
```
It may be worthwhile (if ODD allows it) to move the att.authorized and att.bibl class references into the macro block. Doing so would only allow the attributes in these classes when atomic text is used as content. This restriction would be another way to guiding users toward best practices.

**Comment by pe-ro** [comment]
Here are examples of `<captureMode>` using the last model:
```
<!-- Plain text -->
<captureMode label="Mode of captcha">other text with or without @auth.uri is fine but not elements</captureMode>

<!-- Standard values will be hinted at by oXygen -->
<captureMode label="Mode of captcha" auth.uri="http://id.loc.gov/vocabulary/mcapturestorage/acds">electrical capture, direct
    storage</captureMode>

<!-- Structured text -->
<captureMode>
  <head>Mode of captcha</head>
  <p>Use any mixture of text and <rend rend="bold">textPhaseLike</rend> elements</p>
</captureMode>

<!-- Invalid since @auth.uri is paired with atomic text -->
<captureMode auth.uri="www.foo.bar">
  <head>Mode of captcha</head>
  <p>Use any mixture of text and <rend rend="bold">textPhaseLike</rend> elements</p>
</captureMode>
```

**Reply by pe-ro** [reply]
The last comment above should say "Invalid since `@auth.uri` and structured text can't be used together".
